app/models/concerns/v2/merge_with_alias.rb

```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+module V2
+  # Extended merge method that can harmonize table aliases, mainly for pundit scoping
+  module MergeWithAlias
+    extend ActiveSupport::Concern
+
+    def merge_with_alias(other)
+      other = visitor_for_merge(aliases_for_merge).accept(other.where_clause.ast)
+      other.to_sql == '' ? self : where(other)
+    end
+
+    private
+
+    def aliases_for_merge
+      aliases = {}
+      # Look up all the defined table aliases in the current scope and feed them into the `aliases` hash
+      visitor = ArelVisitor.new { |node| node.is_a?(Arel::Nodes::TableAlias) && aliases[node.table_name] = node.name }
+      visitor.accept(where_clause.ast)
+      aliases
+    end
+
+    # Iterate over the `other` scope and harmonize its table aliases
+    def visitor_for_merge(aliases)
+      ArelVisitor.new(copy: true) do |node|
+        if node.is_a?(Arel::Table) && aliases.key?(node.name)
+          node.alias(aliases[node.name])
+        elsif node.is_a?(Arel::Nodes::TableAlias) && aliases.key?(node.left.name)
+          node.left.alias(aliases[node.left.name])
+        else
+          node
+        end
+      end
+    end
+
+    # Loose implementation of the BFS visitor from the historical Rails codebase
+    class ArelVisitor < Arel::Visitors::Visitor
+      def initialize(copy: false, &block)
+        super()
+        @block = block
+        @copy = copy
+      end
+
+      private
+
+      # :nocov:
+      def visit(node, *args)
+        node = node.dup if @copy
+        super
+        @block.call(node)
+      rescue TypeError => e
+        raise [e.message, 'You should implement an alias for the missing method'].join(': ')
```

Comment:
```
"... in the format of #{"visit_#{(node.class.name || "").gsub("::", "_")}"}"
```

Reply:
That's already in the exception message.